File: demos/advanced/python/04_transactions.py
"""
04_transactions.py — ACID transactions: commit and rollback

Demonstrates:
  - Manual transaction lifecycle (begin → operations → commit)
  - Rollback on failure
  - Isolation: changes not visible until commit (illustrative)
"""

from nocturnusai import SyncNocturnusAIClient
from nocturnusai.exceptions import NocturnusAIError

SERVER = "http://localhost:9300"


def main():
    with SyncNocturnusAIClient(SERVER, database="demo-txn") as client:

        print("=== 1. Successful transaction (commit) ===")
        tx_id = client.begin_transaction()
        print(f"  Started transaction: {tx_id}")

        client.assert_fact("account", ["alice", "1000"], transaction_id=tx_id)
        client.assert_fact("account", ["bob", "500"], transaction_id=tx_id)
        client.commit_transaction(tx_id)
        print("  Committed: account(alice,1000), account(bob,500)")

        accounts = client.query("account", ["?name", "?balance"])
        print(f"  Visible after commit: {len(accounts)} account(s)")
        for a in accounts:
            print(f"    account({', '.join(a.args)})")

        print("\n=== 2. Failed transaction (rollback) ===")
        tx_id2 = client.begin_transaction()
        print(f"  Started transaction: {tx_id2}")

        # Perform writes inside transaction
        client.assert_fact("account", ["charlie", "9999"], transaction_id=tx_id2)

        # Simulate detecting an error before commit
        try:
            raise ValueError("Validation failed — rolling back")
        except ValueError as e:
            print(f"  Error detected: {e}")
            client.rollback_transaction(tx_id2)
            print("  Rolled back transaction")

        charlie = client.query("account", ["charlie", "?bal"])
        print(f"  account(charlie,...) after rollback: {len(charlie)} result(s) (expected 0)")

        print("\n=== 3. Retract inside a transaction ===")
        tx_id3 = client.begin_transaction()
        client.retract("account", ["bob", "500"], transaction_id=tx_id3)
        client.commit_transaction(tx_id3)
        print("  Committed retraction of account(bob,500)")

        accounts_after = client.query("account", ["?name", "?balance"])
        print(f"  Remaining accounts: {len(accounts_after)}")
        for a in accounts_after:
            print(f"    account({', '.join(a.args)})")

        print("\nDone.")


if __name__ == "__main__":
    main()

File: demos/advanced/typescript/04_transactions.ts
/**
 * 04_transactions.ts — ACID transactions: commit and rollback
 *
 * Demonstrates:
 *   - beginTransaction()
 *   - commitTransaction()
 *   - rollbackTransaction()
 *   - Passing transactionId to assertFact / retract
 */

import { NocturnusAIClient } from "nocturnusai-sdk";

const SERVER = "http://localhost:9300";

async function main() {
  const client = new NocturnusAIClient({
    baseUrl: SERVER,
    database: "demo-txn-ts",
  });

  console.log("=== 1. Successful transaction (commit) ===");
  const txId = await client.beginTransaction();
  console.log(`  Started transaction: ${txId}`);

  await client.assertFact("account", ["alice", "1000"], { transactionId: txId });
  await client.assertFact("account", ["bob",   "500"],  { transactionId: txId });
  await client.commitTransaction(txId);
  console.log("  Committed: account(alice,1000), account(bob,500)");

  const accounts = await client.query("account", ["?name", "?balance"]);
  console.log(`  Visible after commit: ${accounts.length} account(s)`);
  for (const a of accounts) {
    console.log(`    account(${a.args.join(", ")})`);
  }

  console.log("\n=== 2. Rollback on simulated error ===");
  const txId2 = await client.beginTransaction();
  console.log(`  Started transaction: ${txId2}`);

  await client.assertFact("account", ["charlie", "9999"], { transactionId: txId2 });

  try {
    throw new Error("Validation failed — rolling back");
  } catch (err) {
    console.log(`  Error detected: ${(err as Error).message}`);
    await client.rollbackTransaction(txId2);
    console.log("  Rolled back transaction");
  }

  const charlie = await client.query("account", ["charlie", "?bal"]);
  console.log(`  account(charlie,...) after rollback: ${charlie.length} result(s) (expected 0)`);

  console.log("\n=== 3. Retract inside a transaction ===");
  const txId3 = await client.beginTransaction();
  await client.retract("account", ["bob", "500"], { transactionId: txId3 });
  await client.commitTransaction(txId3);
  console.log("  Committed retraction of account(bob,500)");

  const remaining = await client.query("account", ["?name", "?balance"]);
  console.log(`  Remaining accounts: ${remaining.length}`);
  for (const a of remaining) {
    console.log(`    account(${a.args.join(", ")})`);
  }

  console.log("\nDone.");
}

main().catch(console.error);
